streaming/base/converters/dataframe_to_mds.py

```diff
@@ -0,0 +1,312 @@
+
+    This method orchestrates the conversion of a spark dataframe into MDS format by
+    processing the input data, applying a user-defined iterable function if
+    provided, and writing the results to MDS-compatible format. The converted data is saved to mds_path.
+
+    Args:
+        dataframe (pyspark.sql.DataFrame): A DataFrame containing Delta Lake data.
+        merge_index (bool): Whether to merge MDS index files. Defaults to ``True``.
+        mds_kwargs (dict): Refer to https://docs.mosaicml.com/projects/streaming/en/stable/api_reference/generated/streaming.MDSWriter.html
+        udf_iterable (Callable or None): A user-defined function that returns an iterable over the dataframe. udf_kwargs is the k-v args for the method. Defaults to ``None``.
+        udf_kwargs (Dict): Additional keyword arguments to pass to the pandas processing
+            function if provided. Defaults to an empty dictionary.
+
+    Return:
+        mds_path (str or (str,str)): actual local and remote path were used
+        fail_count (int): number of records failed to be converted
+
+    Note:
+        - The method creates a SparkSession if not already available.
+        - The 'udf_kwargs' dictionaries can be used to pass additional
+          keyword arguments to the udf_iterable.
+        - If udf_iterable is set, schema check will be skipped because the user defined iterable can create new columns. User must make sure they provide correct mds_kwargs[columns]
+    """
+
+    def write_mds(iterator: Iterable):
+
+        context = TaskContext.get()
+
+        if context is not None:
+            id = context.taskAttemptId()
+        else:
+            raise RuntimeError('TaskContext.get() returns None')
+
+        if isinstance(mds_path, str):  # local
+            output = os.path.join(mds_path, f'{id}')
+            out_file_path = output
+        else:
+            output = (os.path.join(mds_path[0], f'{id}'), os.path.join(mds_path[1], f'{id}'))
+            out_file_path = output[0]
+
+        if mds_kwargs:
+            kwargs = mds_kwargs.copy()
+            kwargs['out'] = output
+        else:
+            kwargs = {}
+
+        if merge_index:
+            kwargs['keep_local'] = True  # need to keep local to do merge
+
+        count = 0
+
+        with MDSWriter(**kwargs) as mds_writer:
+            for pdf in iterator:
+                if udf_iterable is not None:
+                    records = udf_iterable(pdf, **udf_kwargs or {})
+                else:
+                    records = pdf.to_dict('records')
+                assert is_iterable(
+                    records
+                ), f'pandas_processing_fn needs to return an iterable instead of a {type(records)}'
+
+                for sample in records:
+                    try:
+                        mds_writer.write(sample)
+                    except:
+                        raise RuntimeError(f'failed to write sample: {sample}')
+                        count += 1
```

Comment:
I just realized that while my conversation was marked resolved, nothing actually changed here. The count +=1 is unreachable here now so the error counting is kinda useless.